apps/prairielearn/elements/pl-number-input/pl-number-input.py

```diff
@@ -372,68 +386,83 @@ def grade(element_html, data):
-        atol = pl.get_float_attrib(element, "atol", ATOL_DEFAULT)
-        correct = pl.is_correct_scalar_ra(a_sub, a_tru, rtol, atol)
-    elif comparison == "sigfig":
-        digits = pl.get_integer_attrib(element, "digits", DIGITS_DEFAULT)
-        correct = pl.is_correct_scalar_sf(a_sub, a_tru, digits)
-    elif comparison == "decdig":
-        digits = pl.get_integer_attrib(element, "digits", DIGITS_DEFAULT)
-        correct = pl.is_correct_scalar_dd(a_sub, a_tru, digits)
-    else:
-        raise ValueError('method of comparison "%s" is not valid' % comparison)
+    comparison = pl.get_enum_attrib(
+        element, "comparison", ComparisonType, COMPARISON_DEFAULT
+    )
 
-    if correct:
-        data["partial_scores"][name] = {"score": 1, "weight": weight}
-    else:
-        data["partial_scores"][name] = {"score": 0, "weight": weight}
+    def grade_function(
+        a_sub: Union[str, dict[str, Any]]
+    ) -> tuple[bool, Union[str, None]]:
+        nonlocal a_tru
+        # If submitted answer is in a format generated by pl.to_json, convert it
+        # back to a standard type (otherwise, do nothing)
+        a_sub_parsed: Any = pl.from_json(a_sub)
+        feedback = ""
+        is_correct = None
+
+        # Cast both submitted and true answers as np.float64, because...
+        #
+        #   If the method of comparison is relabs (i.e., using relative and
+        #   absolute tolerance) then np.allclose is applied to check if the
+        #   submitted and true answers are the same. If either answer is an
+        #   integer outside the range of int64...
+        #
+        #       https://docs.scipy.org/doc/numpy-1.13.0/user/basics.types.html
+        #
+        #   ...then numpy throws this error:
+        #
+        #       TypeError: ufunc 'isfinite' not supported for the input types, and
+        #       the inputs could not be safely coerced to any supported types
+        #       according to the casting rule ''safe''
+        #
+        #   Casting as np.float64 avoids this error. This is reasonable in any case,
+        #   because <pl-number-input> accepts double-precision floats, not ints.
+        #
+        if np.iscomplexobj(a_sub_parsed) or np.iscomplexobj(a_tru):
+            a_sub_converted = np.complex128(a_sub_parsed)
+            a_tru_converted = np.complex128(a_tru)
+        else:
+            a_sub_converted = np.float64(a_sub_parsed)
+            a_tru_converted = np.float64(a_tru)
+
+        # Compare submitted answer with true answer
+        if comparison is ComparisonType.RELABS:
+            rtol = pl.get_float_attrib(element, "rtol", RTOL_DEFAULT)
+            atol = pl.get_float_attrib(element, "atol", ATOL_DEFAULT)
 
+            a_sub_precision = get_string_precision(str(a_sub))
+            is_correct = pl.is_correct_scalar_ra(
+                a_sub_converted, a_tru_converted, rtol, atol
+            )
+            if not is_correct and (a_sub_precision > rtol):
```

Comment:
Nit: I know we decided against adding the precision warnings for the other grading modes, but might be nice to put a TODO to add that in with a link to this PR in this grading function.

Reply:
Added in [d5c5926](https://github.com/PrairieLearn/PrairieLearn/pull/7934/commits/d5c59269ada670bda3ebf158c0d4557f3dc579e1).

Reply:
I will also open an issue for this after this PR gets merged. @Frozemint